### PR TITLE
Feat: extend apply-once to allow the specified path configuration to drift

### PR DESCRIPTION
**Pull request by fourierr**
Signed-off-by: Xiangbo Ma <maxiangboo@cmbchina.com>

### Description of your changes

Fixes https://github.com/kubevela/kubevela/issues/3840
Extend apply-once to allow the specified path configuration to drift.
There is an example as follow :
```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: fourierapp03
  namespace: fourier
spec:
  components:
    - name: fourierapp03-comp-01
      type: webservice
      properties:
        image: busybox:latest
        imagePullPolicy: IfNotPresent
        cmd: [ "sleep", "10000" ]
        cpu: "0.1"
        mem: "100Mi"
      traits:
        - type: scaler
          properties:
            replicas: 1
  policies:
    - name: fourierapp03-topology-01
      type: topology
      properties:
        clusters: [ "local" ]
        namespace: fourier01
    - name: fourierapp03-topology-02
      type: topology
      properties:
        clusters: [ "local" ]
        namespace: fourier02
    - name: apply-once-01
      type: apply-once
      properties:
        enable: true
        rules:
          - selector:
              componentNames: [ "fourierapp03-comp-01" ]
              resourceTypes: [ "Deployment" ]
            strategy:
              path: [ "*" ]

  workflow:
    steps:
      - type: deploy
        name: fourierapp03-deploy-01
        properties:
          policies: [ "fourierapp03-topology-01" ]
      - type: deploy
        name: fourierapp03-deploy-02
        properties:
          policies: [ "fourierapp03-topology-02" ]
```

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer
ping @wonderflow @Somefive ,please review it.

### CODE REVIEW

File: apis/core.oam.dev/v1alpha1/applyonce_types.go [hunk 1]
```diff
@@ -16,12 +16,51 @@ limitations under the License.
...
 const (
 	// ApplyOncePolicyType refers to the type of configuration drift policy
 	ApplyOncePolicyType = "apply-once"
 )
 
 // ApplyOncePolicySpec defines the spec of preventing configuration drift
 type ApplyOncePolicySpec struct {
-	Enable bool `json:"enable"`
+	Enable bool                  `json:"enable"`
+	Rules  []ApplyOncePolicyRule `json:"rules"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha1/applyonce_types.go:31]: optional fields

File: apis/core.oam.dev/v1alpha1/applyonce_types.go [hunk 2]
```diff
@@ -16,12 +16,51 @@ limitations under the License.
...
+}
+
+// ApplyOncePolicyRule defines a single apply-once policy rule
+type ApplyOncePolicyRule struct {
+	Selector ResourcePolicyRuleSelector `json:"selector"`
+	Strategy ApplyOnceStrategy          `json:"strategy"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha1/applyonce_types.go:37]: optional fields

File: apis/core.oam.dev/v1alpha1/applyonce_types.go [hunk 3]
```diff
@@ -16,12 +16,51 @@ limitations under the License.
...
 // ApplyOncePolicySpec defines the spec of preventing configuration drift
 type ApplyOncePolicySpec struct {
-	Enable bool `json:"enable"`
+	Enable bool                  `json:"enable"`
+	Rules  []ApplyOncePolicyRule `json:"rules"`
+}
+
+// ApplyOncePolicyRule defines a single apply-once policy rule
+type ApplyOncePolicyRule struct {
+	Selector ResourcePolicyRuleSelector `json:"selector"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha1/applyonce_types.go:36]:
```suggestion
	Selector *ResourcePolicyRuleSelector `json:"selectors, omitempty"`
```

optional fields

File: apis/core.oam.dev/v1alpha1/applyonce_types.go (hunk 2 above)
Comment (wonderflow) [reply apis/core.oam.dev/v1alpha1/applyonce_types.go:37]: `*ApplyOnceStrategy` use pointer seems be better

File: apis/core.oam.dev/v1alpha1/applyonce_types.go [hunk 4]
```diff
@@ -16,12 +16,51 @@ limitations under the License.
...
+
+// ApplyOnceStrategy the strategy for resource path to allow configuration drift
+type ApplyOnceStrategy struct {
+	// Path the specified path that allow configuration drift
+	// like 'spec.template.spec.containers[0].resources' and '*' means the whole target allow configuration drift
+	Path []string `json:"path"`
+}
+
+// FindStrategy find apply-once strategy for target resource
+func (in ApplyOncePolicySpec) FindStrategy(manifest *unstructured.Unstructured) *ApplyOnceStrategy {
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha1/applyonce_types.go:48]:
```suggestion
func (in *ApplyOncePolicySpec) FindStrategy(manifest *unstructured.Unstructured) *ApplyOnceStrategy {
```

Comment (wonderflow) [reply apis/core.oam.dev/v1alpha1/applyonce_types.go:48]: FindStrategy => ExtractStrategy 

Will that be more clear?

File: pkg/resourcekeeper/statekeep.go [hunk 5]
```diff
@@ -57,6 +63,13 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
 						return errors.Wrapf(err, "failed to decode resource %s from resourcetracker", mr.ResourceKey())
 					}
 					applyCtx := multicluster.ContextWithClusterName(ctx, mr.Cluster)
+					applyOncePath := h.applyOncePolicy.FindStrategy(manifest)
+					if applyOncePath != nil {
+						manifest, err = applyOnceManifest(applyCtx, h, manifest, applyOncePath)
+						if err != nil {
+							return errors.Wrapf(err, "failed to apply once resource %s from resourcetracker %s", mr.ResourceKey(), rt.Name)
+						}
+					}
```
Comment (wonderflow) [inline pkg/resourcekeeper/statekeep.go:72]: let's merge these lines into a common function like:

```
manifest, err := ApplyStrategies(applyCtx, h, manifest)
```

File: pkg/resourcekeeper/statekeep.go [hunk 6]
```diff
@@ -67,3 +80,160 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
...
+	}
+
+	return manifest, nil
+}
+
+// getNestedValueWithSlice get the value of unstructured.Unstructured via the path, like spec.template.spec.containers[0].image
+func getNestedValueWithSlice(manifest *unstructured.Unstructured, path string) (interface{}, error) {
+label:
+	prefix, index, suffix := splitPathWithIndex(path)
+	// 是 slice
```
Comment (wonderflow) [inline pkg/resourcekeeper/statekeep.go:117]: don't use Chinese in the code

File: pkg/resourcekeeper/statekeep.go [hunk 7]
```diff
@@ -67,3 +80,160 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
...
+			return nil, err
+		}
+		if !found || len(nestedSlice) <= index {
+			return nil, errors.Errorf("the path(%s) was not found ", path)
+		}
+		manifest.Object = nestedSlice[index].(map[string]interface{})
+		path = suffix
+		goto label
+	}
+	// 不是 slice
```
Comment (wonderflow) [inline pkg/resourcekeeper/statekeep.go:131]: the same

File: pkg/resourcekeeper/statekeep.go [hunk 8]
```diff
@@ -67,3 +80,160 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
...
+		prefix = prefixSplit[0]
+		suffixSplit := strings.SplitN(prefixSplit[1], "].", 2)
+		index = cast.ToInt(suffixSplit[0])
+		suffix = suffixSplit[1]
+		return prefix, index, suffix
+	}
+	return path, -1, ""
+}
+
+func toUnstructured(in map[string]interface{}) (*unstructured.Unstructured, error) {
```
Comment (wonderflow) [inline pkg/resourcekeeper/statekeep.go:228]: Maybe reuse this function: `Object2Unstructured`

Or write something similar in the common library.

File: pkg/resourcekeeper/statekeep.go [hunk 9]
```diff
@@ -67,3 +80,160 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
...
+		if !found {
+			return nil, errors.Errorf("the path(%s) was not found ", path)
+		}
+		return nestedField, nil
+	}
+	return nil, nil
+}
+
+// setNestedValueWithSlice set the value to unstructured.Unstructured via the path, like spec.template.spec.containers[0].image
+func setNestedValueWithSlice(manifest *unstructured.Unstructured, path string, value interface{}) (*unstructured.Unstructured, error) {
```
Comment (wonderflow) [inline pkg/resourcekeeper/statekeep.go:147]: `github.com/crossplane/crossplane-runtime/pkg/fieldpath`

Please check if this library can help, we have somthing code related with it, you can search with this package.

It works like this:

```
err = fieldpath.Pave(manifest).SetValue(path, value)
```

**Review by wonderflow — CHANGES_REQUESTED** [review]

**Comment by wonderflow** [comment]
Thanks for the great work

File: apis/core.oam.dev/v1alpha1/applyonce_types.go (hunk 4 above)
Comment (fourierr) [reply apis/core.oam.dev/v1alpha1/applyonce_types.go:48]: FindStrategy keep consistent with GarbageCollectPolicySpec.FindStrategy

File: pkg/resourcekeeper/statekeep.go (hunk 5 above)
Comment (fourierr) [reply pkg/resourcekeeper/statekeep.go:72]: ok, i will modify it

File: pkg/resourcekeeper/statekeep.go (hunk 8 above)
Comment (fourierr) [reply pkg/resourcekeeper/statekeep.go:228]: I'll use object2unstructured, which I haven't found before

**Review by wonderflow — APPROVED** [review]
generally LGTM, ping @Somefive

**Comment by Somefive** [comment]
Suggest adding example docs in https://github.com/kubevela/kubevela/tree/master/docs/examples/app-with-policy .

**Comment by fourierr** [comment]
> Suggest adding example docs in https://github.com/kubevela/kubevela/tree/master/docs/examples/app-with-policy .

ok, i will add the example docs later

**Comment by Somefive** [comment]
Generally LGTM, great job! BTW, is there any way to extract the common part of FindStrategy in ApplyOncePolicySpec and GarbageCollectPolicySpec to ResourcePolicyRuleSelector? @fourierr

**Comment by fourierr** [comment]
After our last discussion, I thought about this issue. mainly including the following points，resulting in no shared findstrategy in `applyoncepolicyspec` and `garbagecollectpolicyspec`.
-  `Garbagecollectpolicyspec` includes not only `garbagecollectpolicyrule`, but also `garbagecollectorder`, etc. At the same time, `applyoncepolicyspec` does not only have `applyoncepolicyrule` fields, and which are used by `findstrategy`
-  The `findstrategy` of `garbagecollectpolicyspec` and `applyoncepolicyspec` has completely different logic. `Garbagecollectpolicyspec` will get the corresponding value from the labels, while `applyoncepolicyspec` is from `App.Policy` gets the corresponding value, and the return parameters are also different.

Based on the above considerations, I think that `garbagecollectpolicyspec` and `applyoncepolicyspec` are different, and the different implementation of `findstrategy` is more clear to user while ensuring extendability. @Somefive

**Comment by Somefive** [comment]
> After our last discussion, I thought about this issue. mainly including the following points，resulting in no shared findstrategy in `applyoncepolicyspec` and `garbagecollectpolicyspec`.
> 
> * `Garbagecollectpolicyspec` includes not only `garbagecollectpolicyrule`, but also `garbagecollectorder`, etc. At the same time, `applyoncepolicyspec` does not only have `applyoncepolicyrule` fields, and which are used by `findstrategy`
> * The `findstrategy` of `garbagecollectpolicyspec` and `applyoncepolicyspec` has completely different logic. `Garbagecollectpolicyspec` will get the corresponding value from the labels, while `applyoncepolicyspec` is from `App.Policy` gets the corresponding value, and the return parameters are also different.
> 
> Based on the above considerations, I think that `garbagecollectpolicyspec` and `applyoncepolicyspec` are different, and the different implementation of `findstrategy` is more clear to user while ensuring extendability. @Somefive

OK.

**Review by Somefive — APPROVED** [review]

**Pull request merged**